File: server/internal/paths/paths.go
package paths

import (
	"errors"
	"os"
	"path/filepath"
	"strings"
)

const AppDirName = ".parsar"

func HomeDir() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil || home == "" {
		return "", errors.New("cannot resolve user home directory")
	}
	return home, nil
}

func Root() (string, error) {
	home, err := HomeDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(home, AppDirName), nil
}

func MustBeUserPath(input string) (string, error) {
	if input == "" {
		return "", errors.New("path is required")
	}
	if strings.HasPrefix(input, "~/") {
		home, err := HomeDir()
		if err != nil {
			return "", err
		}
		return filepath.Join(home, strings.TrimPrefix(input, "~/")), nil
	}
	if filepath.IsAbs(input) {
		return filepath.Clean(input), nil
	}
	return "", errors.New("path must be absolute or start with ~/; relative paths are not accepted")
}

File: packages/cli/src/paths.ts
import os from 'node:os'
import path from 'node:path'

export type ParsarPaths = {
  root: string
  config: string
  logs: string
  state: string
  cache: string
  dev: string
}

export function parsarPaths(): ParsarPaths {
  const root = path.join(os.homedir(), '.parsar')
  return {
    root,
    config: path.join(root, 'config'),
    logs: path.join(root, 'logs'),
    state: path.join(root, 'state'),
    cache: path.join(root, 'cache'),
    dev: path.join(root, 'dev'),
  }
}

export function assertAbsoluteOrTilde(input: string): string {
  if (input.startsWith('~/')) {
    return path.join(os.homedir(), input.slice(2))
  }
  if (path.isAbsolute(input)) {
    return path.resolve(input)
  }
  throw new Error('Path must be absolute or start with ~/. Relative paths are not accepted.')
}
